Updates the UX/UI experience of odo deploy and odo dev

cdrage opened the pull request:
<!--
Thank you for opening a PR! Here are some things you need to know before submitting:

1. Please read our developer guideline: https://github.com/redhat-developer/odo/wiki/Dev:-odo-Dev-Guidelines
2. Label this PR accordingly with the '/kind' line
3. Ensure you have written and ran the appropriate tests: https://github.com/redhat-developer/odo/wiki/Dev:-Writing-and-running-tests
4. Read how we approve and LGTM each PR: https://github.com/redhat-developer/odo/wiki/Pull-Requests:-Review-guideline

Documentation:

If you are pushing a change to documentation, please read: https://github.com/redhat-developer/odo/wiki/Documentation:-Contributing
-->

**What type of PR is this:**

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind tests
/kind documentation

Feel free to use other [labels](https://github.com/redhat-developer/odo/labels) as needed. However one of the above labels must be present or the PR will not be reviewed. This instruction is for reviewers as well.
-->

/kind cleanup

**What does this PR do / why we need it:**

Cleans up the experience when using `odo deploy` and `odo dev` by cleaning up the main
output in a more consise manner.

Other changes:
  * Updates the log/status.go output for "title" and "info"
  * When using `-v` it will show the raw image build / pushing output for debugging purposes.

**UPDATES TO ODO DEPLOY:**

[![asciicast](https://asciinema.org/a/Sde9krmi8jCX0XZzA83UXloIB.svg)](https://asciinema.org/a/Sde9krmi8jCX0XZzA83UXloIB)

**UPDATES TO ODO DEV:**

[![asciicast](https://asciinema.org/a/rhAd0tvzMTje0RpfI6ITTfE7b.svg)](https://asciinema.org/a/rhAd0tvzMTje0RpfI6ITTfE7b)



**Which issue(s) this PR fixes:**
<!--
Specifying the issue will automatically close it when this PR is merged
-->

N/A

**PR acceptance criteria:**

- [X] Unit test

- [X] Integration test

- [X] Documentation

**How to test changes / Special notes to the reviewer:**

Run `odo deploy`

Signed-off-by: Charlie Drage <charlie@charliedrage.com>

deboer-tim commented:
Came here just to say that's very pretty. :)

valaparthvi commented on pkg/devfile/image/docker_compatible.go:
```diff
@@ -24,22 +25,39 @@ func NewDockerCompatibleBackend(name string) *DockerCompatibleBackend {
+		if log.IsDebug() {
+			return fmt.Errorf("error running %s command: %w", o.name, err)
+		}
+		return fmt.Errorf("error running %s command: %w\n\n --- OUTPUT --- \n%s\n\n --- ERROR OUTPUT --- \n%s", o.name, err, cmd.Stdout, cmd.Stderr)
```
Can you only print output/error output if they are not empty?
```sh
➜  odo deploy
  __
 /  \__     Deploying the Devfile: nodejs-prj1-api-abhz
 \__/  \    Namespace: default
 /  \__/    CLI Version: v2.5.0
 \__/

↪ Building & Pushing Container: quay.io/unknown-account/myimage
 ✓  Building image locally [1s]
 ✗  Pushing image to container registry [5s]
 ✗  error running podman command: exit status 125

 --- OUTPUT --- 


 --- ERROR OUTPUT --- 
Getting image source signatures
Error: trying to reuse blob sha256:62a747bf1719d2d37fff5670ed40de6900a95743172de1b4434cb019b56f30b4 at destination: checking whether a blob sha256:62a747bf1719d2d37fff5670ed40de6900a95743172de1b4434cb019b56f30b4 exists in quay.io/unknown-account/myimage: unauthorized: authentication required

➜  export PODMAN_CMD=echo
```
The above empty OUTPUT seems a little weird to me.

valaparthvi commented on pkg/log/status.go:
```diff
@@ -265,6 +265,34 @@ func Swarningf(format string, a ...interface{}) string {
+// Title Prints the logo as well as the first line being BLUE (indicator of the command information)
+// the second and third lines are optional and provide information with regards to what is being ran
+func Title(firstLine, secondLine, thirdLine string) {
+	// Output information regarding the odo deploy command before running
```
Is this specific to odo deploy, or do we intend on using it with all the commands?

kadel replied:
@valaparthvi has a good point here. That comment is misleading as this is used in `odo dev` as well.

valaparthvi commented on pkg/deploy/deploy.go:
```diff
@@ -78,7 +79,7 @@ func (o *deployHandler) ApplyKubernetes(kubernetes v1alpha2.Component) error {
 
 	// Deploy the actual Kubernetes component and error out if there's an issue.
-	log.Infof("\nDeploying Kubernetes %s: %s", u.GetKind(), u.GetName())
+	log.Sectionf("Deploying Kubernetes Component: %s", u.GetName())
```
So, we're going to use log pkg in the business layer? I guess it's much easier to deal with here.

valaparthvi reviewed (commented):
I love this idea of showing odo logo and the lines with a command. Great changes!

kadel reviewed (changes requested):
If we update deploy than we need to update dev as well to match it.

Also `odo deploy` was intentionally showing container build log, now you are hiding it. That is not what we want.

cdrage commented:
@kadel @valaparthvi @dharmit 

This PR has been updated to make changes to both `odo dev` and `odo deploy`.

I have added the image / build output back to `odo deploy` (you could access it via `odo deploy -v 4`, but I've gone ahead and made it re-appear in the standard `odo deploy`.

You can find the new previews here: 

[![asciicast](https://asciinema.org/a/rhAd0tvzMTje0RpfI6ITTfE7b.svg)](https://asciinema.org/a/rhAd0tvzMTje0RpfI6ITTfE7b)

[![asciicast](https://asciinema.org/a/Sde9krmi8jCX0XZzA83UXloIB.svg)](https://asciinema.org/a/Sde9krmi8jCX0XZzA83UXloIB)

cdrage commented:
/retest

dharmit commented:
/lgtm

kadel commented on pkg/odo/cli/dev/dev.go:
```diff
@@ -169,10 +170,17 @@ func (o *DevOptions) Run() error {
 
-	fmt.Fprintf(o.out, "Starting your application on cluster in developer mode\n")
+	// Output what the command is doing / information
+	log.Title("Developing the Devfile: "+devfileName,
```
This might be a little missleading. The user is not developing a Devfile, but an application. 

Maybe `Developing using the "+devfileName+" Devfile`?

kadel commented on pkg/odo/cli/dev/dev.go:
```diff
@@ -169,10 +170,17 @@ func (o *DevOptions) Run() error {
+	// Output what the command is doing / information
+	log.Title("Developing the Devfile: "+devfileName,
+		"Namespace: "+namespace,
+		"Version: "+version.VERSION)
```
```suggestion
		"odo version: "+version.VERSION)
```

kadel commented on pkg/odo/cli/deploy/deploy.go:
```diff
@@ -129,9 +131,24 @@ func (o *DeployOptions) initDevfile() error {
+	// Output what the command is doing / information
+	log.Title("Deploying the Devfile: "+devfileName,
+		"Namespace: "+namespace,
+		"CLI Version: "+version.VERSION)
```
```suggestion
		"odo version: "+version.VERSION)
```

kadel replied:
It might be better to create a wrapper function with just `devfileName` and `namespace` parameters and have rest of the strings inside the function to ensure that the `odo dev` and `odo deploy` output is always i sync

kadel commented on pkg/odo/cli/deploy/deploy.go:
```diff
@@ -129,9 +131,24 @@ func (o *DeployOptions) initDevfile() error {
+	namespace := o.GetProject()
+
+	// Output what the command is doing / information
+	log.Title("Deploying the Devfile: "+devfileName,
```
```suggestion
	log.Title("Deploying the application using "+devfileName+" Devfile",
```

kadel commented:
I would consider updating a few messages, but otherwise this looks good

kadel commented:
The arrow in section line doesn't show correctly on Windows

![image](https://user-images.githubusercontent.com/57206/158585341-415ebeb7-2c0f-479e-85b6-5098e9c5b2d4.png)

cdrage commented:
@kadel thanks! gone ahead and updated the PR. Windows now shows a "dash" instead of the arrow. 

I've also updated the output based on your suggestions.

dharmit commented:
/lgtm
